File: src/main/java/gargoyle/netsquares/res/NSContext.java
package gargoyle.netsquares.res;

import gargoyle.netsquares.util.Assert;

import java.awt.Font;
import java.awt.Image;
import java.io.InputStream;
import java.net.URL;
import java.nio.charset.Charset;
import java.util.List;
import java.util.Locale;
import java.util.ResourceBundle;

public interface NSContext {
    default URL find(String name, String suffix) {
        Locale locale = getLocale();
        ResourceBundle.Control control = ResourceBundle.Control.getControl(ResourceBundle.Control.FORMAT_DEFAULT);
        List<Locale> candidateLocales = control.getCandidateLocales(name, locale);
        for (Locale specificLocale : candidateLocales) {
            String bundleName = control.toBundleName(name, specificLocale);
            String resourceName = control.toResourceName(bundleName, suffix);
            URL location = getClassLoader().getResource(resourceName);
            if (location != null) {
                return location;
            }
        }
        throw new IllegalArgumentException(String.format("%s not found", name));
    }

    Locale getLocale();

    ClassLoader getClassLoader();

    Charset getCharset();

    AudioClip loadAudio(String name, String suffix);

    ResourceBundle loadBundle(String name);

    Font loadFont(String name);

    Image loadImage(String name, String suffix);

    Messages loadMessages(String name);

    default InputStream open(String name) {
        InputStream resource = getClassLoader().getResourceAsStream(name);
        Assert.assertNotNull(resource, String.format("%s not found", name));
        return resource;
    }
}

File: src/main/kotlin/gargoyle/netsquares/res/NSContext.kt
package gargoyle.netsquares.res

import java.awt.Font
import java.awt.Image
import java.io.InputStream
import java.net.URL
import java.nio.charset.Charset
import java.util.Locale
import java.util.ResourceBundle

interface NSContext {
    fun find(name: String, suffix: String): URL? {
        val locale = locale
        val control = ResourceBundle.Control.getControl(ResourceBundle.Control.FORMAT_DEFAULT)
        val candidateLocales = control.getCandidateLocales(name, locale)
        for (specificLocale in candidateLocales) {
            val bundleName = control.toBundleName(name, specificLocale)
            val resourceName = control.toResourceName(bundleName, suffix)
            val location = classLoader.getResource(resourceName)
            if (location != null) {
                return location
            }
        }
        throw IllegalArgumentException(String.format("$name not found"))
    }

    val locale: Locale
    val classLoader: ClassLoader
    val charset: Charset
    fun loadAudio(name: String, suffix: String): AudioClip
    fun loadBundle(name: String): ResourceBundle
    fun loadFont(name: String): Font
    fun loadImage(name: String, suffix: String): Image
    fun loadMessages(name: String): Messages
    fun open(name: String): InputStream {
        return requireNotNull(classLoader.getResourceAsStream(name)) { "$name not found" }
    }
}
